tracer/src/Datadog.Trace/Debugger/Conditions/ProbeExpressionEvaluator.cs

```diff
@@ -0,0 +1,151 @@
+
+        public ProbeExpressionEvaluator(string probeId, DebuggerExpression debuggerExpression, EvaluateAt? evaluateAt)
+            : this(probeId, evaluateAt)
+        {
+            AddCondition(debuggerExpression);
+        }
+
+        public ProbeExpressionEvaluator(string probeId, DebuggerExpression[] debuggerExpressions, EvaluateAt? evaluateAt)
+            : this(probeId, evaluateAt)
+        {
+            AddExpressions(debuggerExpressions);
+        }
+
+        private ProbeExpressionEvaluator(string probeId, EvaluateAt? evaluateAt)
+        {
+            _probeId = probeId;
+            EvaluateAt = evaluateAt;
+            _scopeMembers = new List<ScopeMember>();
+        }
+
+        public List<Lazy<ProbeExpressionParser.R2RExpression<string>>> Expressions { get; set; }
+
+        public Lazy<ProbeExpressionParser.R2RExpression<bool>> Condition { get; set; }
+
+        // we can save this as T but it will require native changes to return MethodDebuggerState<T> instead MethodDebuggerState
+        public ScopeMember InvocationTarget { get; private set; }
+
+        public EvaluateAt? EvaluateAt { get; }
+
+        public ReadOnlyCollection<ScopeMember> ScopeMembers => _scopeMembers.AsReadOnly();
+
+        private void AddMember(string name, Type type, object value, ScopeMember.MemberType eleMemberType)
+        {
+            if (!Enum.IsDefined(typeof(ScopeMember.MemberType), eleMemberType))
+            {
+                throw new ArgumentException(nameof(eleMemberType) + "is not a valid enum value");
+            }
+
+            _scopeMembers.Add(new ScopeMember(name, type, value, eleMemberType));
+        }
+
+        public void AddArgument(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Argument);
+        }
+
+        public void AddLocal(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Local);
+        }
+
+        public void AddException(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Exception);
+        }
+
+        public void AddReturn(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Return);
+        }
+
+        public void SetInvocationTarget(string name, Type type, object value)
+        {
+            InvocationTarget = new ScopeMember(name, type, value, ScopeMember.MemberType.This);
+        }
+
+        public bool TryEvaluateCondition(out bool result)
+        {
+            _shouldEvaluate = false;
+            try
+            {
+                result = Condition.Value.Delegate(InvocationTarget, ScopeMembers);
+                return true;
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, $"Failed to evaluate condition for probe {_probeId}. Condition: {Condition.Value.Dsl}");
```

Comment:
Now that this code is generic and isn't only for conditions, this error message is incorrect. Maybe consider handling this exception at the callee, in which case, you can probably get rid of the `_probeId` field and not even pass it into this class?